src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ServiceClientService.java

```diff
@@ -0,0 +1,252 @@
+import ee.ria.xroad.common.identifier.ClientId;
+import ee.ria.xroad.common.identifier.XRoadId;
+
+import lombok.extern.slf4j.Slf4j;
+import org.niis.xroad.restapi.dto.ServiceClientAccessRightDto;
+import org.niis.xroad.restapi.dto.ServiceClientDto;
+import org.niis.xroad.restapi.dto.ServiceClientIdentifierDto;
+import org.niis.xroad.restapi.openapi.ResourceNotFoundException;
+import org.niis.xroad.restapi.repository.ClientRepository;
+import org.niis.xroad.restapi.util.FormatUtils;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Service;
+
+import javax.transaction.Transactional;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Service
+@Transactional
+@PreAuthorize("isAuthenticated()")
+public class ServiceClientService {
+
+    private final ClientRepository clientRepository;
+    private final ServiceService serviceService;
+    private final EndpointService endpointService;
+    private final AccessRightService accessRightService;
+    private final LocalGroupService localGroupService;
+
+    public ServiceClientService(ClientRepository clientRepository, ServiceService serviceService,
+            EndpointService endpointService, AccessRightService accessRightService,
+            LocalGroupService localGroupService) {
+        this.clientRepository = clientRepository;
+        this.serviceService = serviceService;
+        this.endpointService = endpointService;
+        this.accessRightService = accessRightService;
+        this.localGroupService = localGroupService;
+    }
+
+    /**
+     * Get access right holders (serviceClients) by Client (service owner)
+     *
+     * The concept of base endpoint is used to find service level access rights in this method.
+     * Base endpoint is in other words service (code) level endpoint.
+     * Each service has one base endpoint.
+     * Base endpoint has method '*' and path '**'.
+     *
+     * @param ownerId
+     * @return
+     * @throws ClientNotFoundException
+     *
+     */
+    public List<ServiceClientDto> getServiceClientsByClient(ClientId ownerId)
+            throws ClientNotFoundException {
+        ClientType owner = clientRepository.getClient(ownerId);
+        if (owner == null) {
+            throw new ClientNotFoundException("Client " + ownerId.toShortString() + " not found");
+        }
+
+        // Filter just acls that are set to base endpoints so they are on service code level
+        List<AccessRightType> serviceCodeLevelAcls = owner.getAcl().stream()
+                .filter(acl -> acl.getEndpoint().isBaseEndpoint())
+                .collect(Collectors.toList());
+        List<AccessRightType> distinctAccessRightTypes = distinctAccessRightTypeByXroadId(serviceCodeLevelAcls);
+        return accessRightService.mapAccessRightsToServiceClients(owner, distinctAccessRightTypes);
```

Comment:
Just realized one problem. When listing service client(s) for one client, I think we should not return `rights_given_at`

```
    {
        "id": "85",
        "name": "Aston Martin Red Bull Racing",
        "local_group_code": "RBR",
        "service_client_type": "LOCALGROUP",
        "rights_given_at": "2020-04-15T03:46:21.297Z"
    },
```
Reason is that we really do not know when the rights were given. Aston Martin Red Bull Racing is a service client for this client's services, but it can be a service client for service A, B and C. `distinctAccessRightTypeByXroadId` picks AccessRightType given to A, B or C **by random**. It is possible that serviceclient.rights_given_at should not be `2020-04-15T03:46:21.297Z` but maybe `2019-01-01` instead. And it is also possible that returned timestamp changes randomly from request to another.

When listing service client(s) for a client, **rights_given_at should be null** since we do not know it. On the other hand when listing service client(s) the other way (`like /services/{id}/service-clients`) we **do know** the real rights_given_at.

Alternative to nulling it would be to scan through all AccessRightTypes and pick the earliest timestamp, that would then be the correct value. But I think that is unnecessary, as we do not use that in the UI.

Reply:
I found a rather easy way to populate the rights_given_at -information with the earliest timestamp. IMO it's better than mapping all those values to null after they've been first populated (which is the other option)

Reply:
OK, that's nice!